src/components/lib/cos_kernel_api.c

```diff
@@ -689,11 +808,33 @@ cos_hw_cycles_per_usec(hwcap_t hwc)
 { return call_cap_op(hwc, CAPTBL_OP_HW_CYC_USEC, 0, 0, 0, 0); }
 
 void *
-cos_hw_map(struct cos_compinfo *ci, hwcap_t hwc, paddr_t pa)
+cos_hw_map(struct cos_compinfo *ci, hwcap_t hwc, paddr_t pa, unsigned int len)
 {
-	vaddr_t va = __page_bump_valloc(ci);
+	size_t sz;
+	vaddr_t fva, va;
+
+	assert(ci && hwc && pa && len);
+
+	sz   = round_up_to_page(len);
+
+	fva = __page_bump_valloc(ci);
+	if (unlikely(!fva)) return NULL;
+
+	if(call_cap_op(hwc, CAPTBL_OP_HW_MAP, ci->pgtbl_cap, fva, pa, 0)) BUG();
```

Comment:
what am I missing?  Could all of this be simplified with a `do { } while(...);`?  I don't like seeing the redundancy.